PHPLIB-1522: Deprecate info iterators

alcaeus opened the pull request:
PHPLIB-1522

This PR deprecates the `CollectionInfoIterator`, `DatabaseInfoIterator`, and `IndexInfoIterator` interfaces and implementing classes. These will be replaced with generic iterators in 2.0.

In addition to this change, this PR also deprecates `IndexInfo::getNamespace`. The `ns` field that feeds this method has been removed from output in MongoDB 4.4, and keeping the necessary code added in PHPLIB-499 around makes no sense going forward.

alcaeus committed:
PHPLIB-1522: Deprecate info iterators

alcaeus committed:
Remove usages of info iterator interfaces in return types

alcaeus committed:
Deprecate IndexInfo::getNamespace

GromNaN commented on src/Client.php:
```diff
@@ -294,7 +293,7 @@ public function listDatabaseNames(array $options = []): Iterator
      *
      * @see ListDatabases::__construct() for supported options
-     * @return DatabaseInfoIterator
+     * @return Iterator
```
Can't you use the generics syntax here? Otherwise this change is a regression for typing.
```suggestion
     * @return Iterator<DatabaseInfo>
```

GromNaN commented on src/Collection.php:
```diff
@@ -882,7 +881,7 @@ public function insertOne($document, array $options = [])
      *
      * @see ListIndexes::__construct() for supported options
-     * @return IndexInfoIterator
+     * @return Iterator
```
```suggestion
     * @return Iterator<IndexInfo>
```

GromNaN commented on src/Database.php:
```diff
@@ -473,7 +472,7 @@ public function listCollectionNames(array $options = []): Iterator
      *
      * @see ListCollections::__construct() for supported options
-     * @return CollectionInfoIterator
+     * @return Iterator
```
```suggestion
     * @return Iterator<CollectionInfo>
```

GromNaN reviewed (commented):
I would change the assertion on tests also:
https://github.com/mongodb/mongo-php-library/blob/49677cbc93dceb40035877978ab17d57035ec8e3/tests/ClientFunctionalTest.php#L61

alcaeus committed:
Use templated return type

alcaeus commented:
> I would change the assertion on tests also:
> 
> https://github.com/mongodb/mongo-php-library/blob/49677cbc93dceb40035877978ab17d57035ec8e3/tests/ClientFunctionalTest.php#L61

For now I'd leave in the assertion as a safe guard. We can update this in 2.x. I have added the templated return types for Psalm's benefit and updated the baseline.

GromNaN reviewed (approved):
LGTM
